### PR TITLE
scripts: twister: set Python3_EXECUTABLE when calling cmake

### PR DESCRIPTION
**Pull request by mtfurlan**
`west build` sets WEST_PYTHON when calling cmake, but `west twister`
does not
This causes things to not work as expected when using a non-global python

The path code was copied from the scripts/west_commands/build.py

Signed-off-by: Mara Furland <mara@fur.land>

### CODE REVIEW

**Review by pdgendt — COMMENTED** [review]
Fix commit message as described in the [guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#pull-request-guidelines)

**Comment by mtfurlan** [comment]
It's failing tests such as `test_projectbuilder_cmake_assemble_args_single` in `scripts/tests/twister/test_runner.py` cause the cmake args are tested against a hardcoded list

There don't appear to be any similar tests for `west build` to copy, and I don't think hardcoding the string `/opt/hostedtoolcache/Python/3.10.15/x64/bin/python` is acceptable in tests, so how should these tests get fixed?

File: scripts/pylib/twister/twisterlib/runner.py [hunk 1]
```diff
@@ -1179,6 +1180,8 @@ def cmake_assemble_args(extra_args, handler, extra_conf_files, extra_overlay_con
         args_expanded.extend(["-D{}".format(a.replace('"', '\"')) for a in cmake_extra_args])
         args_expanded.extend(["-D{}".format(a.replace('"', '')) for a in args])
 
+        args_expanded.append('-DWEST_PYTHON={}'.format(pathlib.Path(sys.executable).as_posix()))
```
Comment (golowanow) [inline scripts/pylib/twister/twisterlib/runner.py:1183]: @tejlmand, should it be `Python3_EXECUTABLE`  ? #72232
```suggestion
        args_expanded.append('-DPython3_EXECUTABLE={}'.format(pathlib.Path(sys.executable).as_posix()))
```

**Comment by golowanow** [comment]
>..  how should these tests get fixed?

I think the test can be adjusted to compose expected args using the current python3, or mock sys.executable

File: scripts/pylib/twister/twisterlib/runner.py (hunk 1 above)
Comment (tejlmand) [reply scripts/pylib/twister/twisterlib/runner.py:1183]: > @tejlmand, should it be Python3_EXECUTABLE ? #72232

yes it should. `WEST_PYTHON` is intended for `west`.

**Review by tejlmand — CHANGES_REQUESTED** [review]
> It's failing tests such as test_projectbuilder_cmake_assemble_args_single in scripts/tests/twister/test_runner.py cause the cmake args are tested against a hardcoded list

because you should make your adjustment here instead:
https://github.com/zephyrproject-rtos/zephyr/blob/86b666548b97f0fa5f4ec05e9b6f92a2bfe5a4ed/scripts/pylib/twister/twisterlib/runner.py#L343-L350

**Comment by mtfurlan** [comment]
Okay, thanks for the feedback!
I switched to defining Python3_EXECUTABLE in the cmake_args section, and I added the defines to the test runner so the tests pass(at least locally)

**Comment by mtfurlan** [comment]
It looks like [twister blackbox `test_coverage.py` `test_coverage_basedir`](https://github.com/mtfurlan/zephyr/blob/twister-set-west-python/scripts/tests/twister_blackbox/test_coverage.py#L247) is failing because the regex `'{"files": \\[], "gcovr/format_version": ".*"}'` does not match the json `{'gcovr/format_version': '0.10', 'files': []}` which I don't think is related to this change?

**Comment by golowanow** [comment]
> It looks like [twister blackbox `test_coverage.py` `test_coverage_basedir`](https://github.com/mtfurlan/zephyr/blob/twister-set-west-python/scripts/tests/twister_blackbox/test_coverage.py#L247) is failing because the regex `'{"files": \\[], "gcovr/format_version": ".*"}'` does not match the json `{'gcovr/format_version': '0.10', 'files': []}` which I don't think is related to this change?

probably it was lucky random mismatch to the wrong pattern - to fix it with #79578

**Comment by mtfurlan** [comment]
> probably it was lucky random mismatch to the wrong pattern - to fix it with #79578

That patch fixes it for me locally
That test fails the same way locally on the main branch, and the patch fixes it there too

**Review by tejlmand — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by golowanow — APPROVED** [review]

**Pull request merged**